query/raw_result.go

```diff
@@ -212,19 +213,98 @@ func (r *RawResults) IsOrdered() bool {
 }
 
 // LimitReached returns true if we have collected enough raw results.
-// TODO(xichen): Implement this.
 func (r *RawResults) LimitReached() bool {
-	panic("not implemented")
+	return len(r.Data) >= r.Limit
 }
 
 // Add adds a raw result to the collection.
-// TODO(xichen): Implement this.
+// For unordered raw results:
+// - If the results have not reached limit yet, the incoming result is appended at the end.
+// - Otherwise, the incoming result is dropped.
+// For ordered raw results:
+// - If the results have not reached limit yet, the incoming result is added in order.
+// - Otherwise, the incoming result is inserted and the last result is dropped.
 func (r *RawResults) Add(rr RawResult) {
+	if !r.IsOrdered() {
+		if r.LimitReached() {
+			return
+		}
+		if r.Data == nil {
+			r.Data = make([]RawResult, 0, r.Limit)
+		}
+		r.Data = append(r.Data, rr)
+		return
+	}
+
+	// TODO(xichen): We currently don't use `Add` for ordered raw results so punt on the
+	// implementation for now. When we do, a templatized linked list might be a better choice
+	// for the result collection.
 	panic("not implemented")
 }
 
 // AddBatch adds a batch of raw results to the collection.
-// TODO(xichen): Implement this.
+// For unordered raw results, the incoming batch is unsorted:
+// - If the results have not reached limit yet, the incoming results are appended at the end
+//   until the limit is reached, after which the incoming results are dropped.
```

Comment:
Haven't looked but should we add a `TODO` to add short circuiting logic to stop queries from executing once the result limit is reached?

Reply:
Yeah this has been handled in the segment query logic.